File: mobile/ios/KinicApp/Services/SourceCaptureRequestBuilder.swift
// Where: mobile/ios/KinicApp/Services/SourceCaptureRequestBuilder.swift
// What: Builds the native equivalent of wikibrowser source-capture request nodes.
// Why: iOS submissions must preserve the existing worker contract exactly.

import Foundation

enum SourceCaptureRequestBuilder {
    static func request(
        url: URL,
        databaseId: String,
        requestedBy: String,
        requestId: String? = nil,
        now: Date = .now,
        uuid: UUID = UUID(),
        outputLanguage: WikiOutputLanguage = .english,
        captureMetadata: ShareCaptureMetadata? = nil
    ) throws -> SourceCaptureRequest {
        let normalizedURL = try URLNormalizer.normalizedHTTPURL(url)
        let resolvedRequestId: String
        if let requestId {
            resolvedRequestId = try validateRequestId(requestId)
        } else {
            resolvedRequestId = try makeRequestId(now: now, uuid: uuid)
        }
        guard let requestPath = SourceCaptureContract.requestPath(for: resolvedRequestId) else {
            throw SourceCaptureRequestError.invalidRequestId
        }
        let requestedAt = now.formatted(.iso8601)
        let urlText = normalizedURL.absoluteString
        let frontmatter = [
            "---",
            "kind: kinic.source_capture_request",
            "schema_version: 1",
            "status: queued",
            "url: \(jsonString(urlText))",
            "requested_by: \(jsonString(requestedBy))",
            "requested_at: \(jsonString(requestedAt))",
            "output_language: \(jsonString(outputLanguage.rawValue))",
            "claimed_at: null",
            "source_path: null",
            "target_path: null",
            "finished_at: null",
            "error: null"
        ]
        let metadataFrontmatter = captureMetadataFrontmatter(captureMetadata)
        let body = [
            "---",
            "",
            "# Source Capture Request",
            ""
        ] + captureMetadataBody(captureMetadata)
        let content = (frontmatter + metadataFrontmatter + body).joined(separator: "\n")
        var metadataPayload = [
            "request_type": "source_capture",
            "url": urlText,
            "output_language": outputLanguage.rawValue
        ]
        appendCaptureMetadata(captureMetadata, to: &metadataPayload)
        let metadata = try JSONSerialization.data(
            withJSONObject: metadataPayload,
            options: [.sortedKeys]
        )
        let metadataJson = String(data: metadata, encoding: .utf8) ?? "{}"
        return SourceCaptureRequest(
            databaseId: databaseId,
            requestId: resolvedRequestId,
            requestPath: requestPath,
            content: content,
            metadataJson: metadataJson,
            normalizedURL: normalizedURL,
            outputLanguage: outputLanguage
        )
    }

    static func makeRequestId(now: Date = .now, uuid: UUID = UUID()) throws -> String {
        try safeRequestId(timeMs: milliseconds(now), uuid: uuid.uuidString.lowercased())
    }

    static func validateRequestId(_ requestId: String) throws -> String {
        guard isSafeStorageSegment(requestId, maxLength: 128) else {
            throw SourceCaptureRequestError.invalidRequestId
        }
        return requestId
    }

    static func safeRequestId(timeMs: Int64, uuid: String) throws -> String {
        let suffix = uuid.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isSafeStorageSegment(suffix, maxLength: 96) else {
            throw SourceCaptureRequestError.invalidRequestId
        }
        let requestId = "\(timeMs)-\(suffix)"
        guard isSafeStorageSegment(requestId, maxLength: 128) else {
            throw SourceCaptureRequestError.invalidRequestId
        }
        return requestId
    }

    private static func milliseconds(_ date: Date) -> Int64 {
        Int64((date.timeIntervalSince1970 * 1_000).rounded(.down))
    }

    private static func captureMetadataFrontmatter(_ metadata: ShareCaptureMetadata?) -> [String] {
        guard let metadata, metadata.hasContent else {
            return []
        }
        var lines = [
            "shared_metadata_source: \(jsonString(metadata.source))",
            "shared_metadata_fetched_at: \(jsonString(metadata.fetchedAt.formatted(.iso8601)))"
        ]
        if let title = metadata.title {
            lines.append("shared_title: \(jsonString(title))")
        }
        if let description = metadata.description {
            lines.append("shared_description: \(jsonString(description))")
        }
        if let imageURL = metadata.imageURL {
            lines.append("shared_image_url: \(jsonString(imageURL.absoluteString))")
        }
        return lines
    }

    private static func captureMetadataBody(_ metadata: ShareCaptureMetadata?) -> [String] {
        guard let metadata, metadata.hasContent else {
            return []
        }
        var lines = [
            "## Shared Metadata",
            "",
            "Source: \(metadata.source)",
            "Fetched at: \(metadata.fetchedAt.formatted(.iso8601))"
        ]
        if let title = metadata.title {
            lines.append("Title: \(title)")
        }
        if let imageURL = metadata.imageURL {
            lines.append("Image: \(imageURL.absoluteString)")
        }
        if let description = metadata.description {
            lines.append("")
            lines.append("### Description")
            lines.append("")
            lines.append(description)
        }
        lines.append("")
        return lines
    }

    private static func appendCaptureMetadata(
        _ metadata: ShareCaptureMetadata?,
        to payload: inout [String: String]
    ) {
        guard let metadata, metadata.hasContent else {
            return
        }
        payload["shared_metadata_source"] = metadata.source
        payload["shared_metadata_fetched_at"] = metadata.fetchedAt.formatted(.iso8601)
        if let title = metadata.title {
            payload["shared_title"] = title
        }
        if let description = metadata.description {
            payload["shared_description"] = description
        }
        if let imageURL = metadata.imageURL {
            payload["shared_image_url"] = imageURL.absoluteString
        }
    }

    static func isSafeStorageSegment(_ value: String, maxLength: Int = 128) -> Bool {
        maxLength == 128
            ? SourceCaptureContract.isSafeRequestId(value)
            : value.utf8.count <= maxLength && SourceCaptureContract.isSafeRequestId(value)
    }

    private static func jsonString(_ value: String) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: value, options: [.fragmentsAllowed]),
              let text = String(data: data, encoding: .utf8) else {
            return "\"\""
        }
        return text
    }
}

File: mobile/android/app/src/main/java/xyz/kinic/android/SourceCaptureRequestBuilder.kt
// Where: mobile/android/app/src/main/java/xyz/kinic/android/SourceCaptureRequestBuilder.kt
// What: Builds Android source capture request nodes.
// Why: Shared links must preserve the existing worker contract exactly.

package xyz.kinic.android

import java.net.URI
import java.time.Instant
import java.util.UUID

object SourceCaptureRequestBuilder {
    private val safeFirst = Regex("^[A-Za-z0-9]")
    private val safeAll = Regex("^[A-Za-z0-9._-]+$")

    fun request(
        url: URI,
        databaseId: String,
        requestedBy: String,
        requestId: String? = null,
        now: Instant = Instant.now(),
        uuid: UUID = UUID.randomUUID(),
        captureMetadata: ShareCaptureMetadata? = null,
        outputLanguage: WikiOutputLanguage = WikiOutputLanguage.ENGLISH,
    ): SourceCaptureRequest {
        val normalizedUrl = URLNormalizer.normalizedHttpUrl(url)
        val resolvedRequestId = requestId?.let(::validateRequestId) ?: makeRequestId(now, uuid)
        val requestPath = "/Sources/source-capture-requests/$resolvedRequestId.md"
        val urlText = normalizedUrl.toString()
        val cleanedMetadata = captureMetadata?.cleaned()

        val frontmatter = mutableListOf(
            "---",
            "kind: kinic.source_capture_request",
            "schema_version: 1",
            "status: queued",
            "url: ${jsonString(urlText)}",
            "requested_by: ${jsonString(requestedBy)}",
            "requested_at: ${jsonString(now.toString())}",
            "output_language: ${jsonString(outputLanguage.code)}",
            "claimed_at: null",
            "source_path: null",
            "target_path: null",
            "finished_at: null",
            "error: null",
        )
        frontmatter += captureMetadataFrontmatter(cleanedMetadata)
        val body = listOf("---", "", "# Source Capture Request", "") +
            captureMetadataBody(cleanedMetadata)

        val metadataPayload = mutableMapOf(
            "request_type" to "source_capture",
            "url" to urlText,
            "output_language" to outputLanguage.code,
        )
        appendCaptureMetadata(cleanedMetadata, metadataPayload)

        return SourceCaptureRequest(
            databaseId = databaseId,
            requestId = resolvedRequestId,
            requestPath = requestPath,
            content = (frontmatter + body).joinToString("\n"),
            metadataJson = jsonObjectSorted(metadataPayload),
            normalizedUrl = normalizedUrl,
            outputLanguage = outputLanguage,
        )
    }

    fun makeRequestId(now: Instant = Instant.now(), uuid: UUID = UUID.randomUUID()): String =
        safeRequestId(now.toEpochMilli(), uuid.toString().lowercase())

    fun validateRequestId(requestId: String): String {
        if (!isSafeStorageSegment(requestId)) {
            throw SourceCaptureRequestException.InvalidRequestId
        }
        return requestId
    }

    fun safeRequestId(timeMs: Long, uuid: String): String {
        val suffix = uuid.trim()
        if (!isSafeStorageSegment(suffix, maxLength = 96)) {
            throw SourceCaptureRequestException.InvalidRequestId
        }
        val requestId = "$timeMs-$suffix"
        if (!isSafeStorageSegment(requestId)) {
            throw SourceCaptureRequestException.InvalidRequestId
        }
        return requestId
    }

    fun isSafeStorageSegment(value: String, maxLength: Int = 128): Boolean =
        value.length <= maxLength &&
            value != "." &&
            value != ".." &&
            !value.contains("..") &&
            safeFirst.containsMatchIn(value) &&
            safeAll.matches(value)

    private fun captureMetadataFrontmatter(metadata: ShareCaptureMetadata?): List<String> {
        if (metadata?.hasContent != true) return emptyList()
        val lines = mutableListOf(
            "shared_metadata_source: ${jsonString(metadata.source)}",
            "shared_metadata_fetched_at: ${jsonString(metadata.fetchedAt.toString())}",
        )
        metadata.title?.let { lines += "shared_title: ${jsonString(it)}" }
        metadata.description?.let { lines += "shared_description: ${jsonString(it)}" }
        metadata.imageUrl?.let { lines += "shared_image_url: ${jsonString(it.toString())}" }
        return lines
    }

    private fun captureMetadataBody(metadata: ShareCaptureMetadata?): List<String> {
        if (metadata?.hasContent != true) return emptyList()
        val lines = mutableListOf(
            "## Shared Metadata",
            "",
            "Source: ${metadata.source}",
            "Fetched at: ${metadata.fetchedAt}",
        )
        metadata.title?.let { lines += "Title: $it" }
        metadata.imageUrl?.let { lines += "Image: $it" }
        metadata.description?.let {
            lines += ""
            lines += "### Description"
            lines += ""
            lines += it
        }
        lines += ""
        return lines
    }

    private fun appendCaptureMetadata(metadata: ShareCaptureMetadata?, payload: MutableMap<String, String>) {
        if (metadata?.hasContent != true) return
        payload["shared_metadata_source"] = metadata.source
        payload["shared_metadata_fetched_at"] = metadata.fetchedAt.toString()
        metadata.title?.let { payload["shared_title"] = it }
        metadata.description?.let { payload["shared_description"] = it }
        metadata.imageUrl?.let { payload["shared_image_url"] = it.toString() }
    }
}

sealed class SourceCaptureRequestException(message: String) : IllegalArgumentException(message) {
    data object InvalidRequestId : SourceCaptureRequestException("invalid request id")
}
